src/Functions/bitSlice.cpp

```diff
@@ -0,0 +1,476 @@
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/GatherUtils/Algorithms.h>
+#include <Functions/GatherUtils/GatherUtils.h>
+#include <Functions/GatherUtils/Sinks.h>
+#include <Functions/GatherUtils/Slices.h>
+#include <Functions/GatherUtils/Sources.h>
+#include <Functions/IFunction.h>
+#include <IO/WriteHelpers.h>
+
+namespace DB
+{
+using namespace GatherUtils;
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int ZERO_ARRAY_OR_TUPLE_INDEX;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+class FunctionBitSlice : public IFunction
+{
+    const UInt8 word_size = 8;
+
+public:
+    static constexpr auto name = "bitSlice";
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionBitSlice>(); }
+
+    String getName() const override { return name; }
+
+    bool isVariadic() const override { return true; }
+    size_t getNumberOfArguments() const override { return 0; }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        const size_t number_of_arguments = arguments.size();
+
+        if (number_of_arguments < 2 || number_of_arguments > 3)
+            throw Exception(
+                "Number of arguments for function " + getName() + " doesn't match: passed " + toString(number_of_arguments)
+                    + ", should be 2 or 3",
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        if (!isString(arguments[0]) && !isStringOrFixedString(arguments[0]))
+            throw Exception(
+                "Illegal type " + arguments[0]->getName() + " of argument of function " + getName(), ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        if (arguments[0]->onlyNull())
+            return arguments[0];
+
+        if (!isNativeNumber(arguments[1]))
+            throw Exception(
+                "Illegal type " + arguments[1]->getName() + " of second argument of function " + getName(),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        if (number_of_arguments == 3 && !isNativeNumber(arguments[2]))
+            throw Exception(
+                "Illegal type " + arguments[2]->getName() + " of second argument of function " + getName(),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        return std::make_shared<DataTypeString>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        size_t number_of_arguments = arguments.size();
+
+        ColumnPtr column_string = arguments[0].column;
+        ColumnPtr column_start = arguments[1].column;
+        ColumnPtr column_length;
+
+        if (number_of_arguments == 3)
+            column_length = arguments[2].column;
```

Comment:
Let's assign `column_length` and `column_length_const` under same if `if (number_of_arguments == 3)` below, because this variables connected

Reply:
OK